// Copyright (c) 2016, myjfm(mwxjmmyjfm@gmail.com).  All rights reserved.
// This source code is licensed under the BSD-style license found in the
// LICENSE file in the root directory of this source tree. An additional grant
// of patent rights can be found in the PATENTS file in the same directory.
//
syntax = "proto3";

package uranium.common;

enum TableType {
  KV      = 0;
  LIST    = 1;
  HASH    = 2;
  SET     = 3;
  SCHEMA  = 4;
}

message TableName {
  string name = 1;
}

enum Status {
  OK                    = 0;
  INTERNAL_ERROR        = 1;
  TABLE_NOT_FOUND       = 2;
  KEY_NOT_FOUND         = 3;
  TABLE_ALREADY_EXISTS  = 4;
  KEY_ALREADY_EXISTS    = 5;
  INVALID_ARGUMENT      = 6;
  OUT_OF_RANGE          = 7;
}

message Result {
  Status status   = 1;
  string message  = 2;
}

message ColumnName {
  string name = 1;
}

enum ColumnType {
  BYTE      = 0;
  SHORT     = 1;
  INT       = 2;
  LONG      = 3;
  FLOAT     = 4;
  DOUBLE    = 5;
  STRING    = 6;
  DATE      = 7;
  DATETIME  = 8;
}

message ColumnValue {
  oneof value {
    sint64 integer  = 1;
    float flt       = 2;
    double dbl      = 3;
    bytes other     = 4;
  }
}

message ColumnDefination {
  ColumnName column_name    = 1;
  ColumnType column_type    = 2;
  ColumnValue default_value = 3;
  bool not_null             = 4;
}

